### PR TITLE
namespace run_targets by subproject

### PR DESCRIPTION
**Pull request by martingkelly**
namespace run_targets by subproject
    
Currently, run_target does not get namespaced for each subproject,
unlike executable and others. This means that two subprojects sharing
the same run_target name cause meson to crash.
    
Fix this by moving the subproject namespacing logic from the BuildTarget
class to the Target class.

### CODE REVIEW

File: mesonbuild/modules/gnome.py [hunk 1]
```diff
@@ -215,7 +215,7 @@ def compile_resources(self, state, args, kwargs):
             h_kwargs['install'] = install_header
             h_kwargs['install_dir'] = kwargs.get('install_dir',
                                                  state.environment.coredata.get_builtin_option('includedir'))
-        target_h = GResourceHeaderTarget(args[0] + '_h', state.subdir, h_kwargs)
+        target_h = GResourceHeaderTarget(args[0] + '_h', state.subdirs, state.subproject, h_kwargs)
```
Comment (nirbheek) [inline mesonbuild/modules/gnome.py]: `s/subdirs/subdir` this is causing the CI failure.

**Comment by martingkelly** [comment]
Looks like the errors found have to do with a `pip3 install` fail rather than the code itself.

**Comment by nirbheek** [comment]
https://github.com/mesonbuild/meson/pull/2294

**Comment by martingkelly** [comment]
@nirbheek looks like @jpakkane removed coverage checking entirely. Given that, is this ready to merge, or would you like to do another test run first with coverage removed?

**Comment by martingkelly** [comment]
Rebasing to force a CI re-run, and for good hygiene.

**Comment by nirbheek** [comment]
Yes, we can merge this once the CI passes. Coverage is anyway a bit broken due to reasons, so it doesn't matter for this PR.

**Comment by martingkelly** [comment]
Excellent, thanks!

**Comment by jpakkane** [comment]
This implementation names the target with a hyphen, .e.g `foo-nop`. That means that if there is a top level target called `foo-nop`, there will be a clash. Maybe we should use `:` instead because that is what we do elsewhere. `:` is not a shell metacharacter in sh at least, what about Windows shell?

**Comment by martingkelly** [comment]
(ping) @nirbheek @jpakkane

**Comment by nirbheek** [comment]
In other places, we assume that '@' is an infrequently-used directory and target name; for instance to `foo@exe/` for an `executable` called `foo`. Perhaps you can do the same here.

**Comment by nirbheek** [comment]
Correction, we use `@@`. See `BuildTarget.get_id()` for instance.

**Comment by martingkelly** [comment]
OK, I revised it to use `@@` instead of `-`. Let me know if there's anything else blocking merge.

**Comment by martingkelly** [comment]
Looks like the appveyor build failure is unrelated.

**Comment by martingkelly** [comment]
Rebased to fix merge conflicts and pushed again.

**Comment by martingkelly** [comment]
Again the test failures appear unrelated. @nirbheek @jpakkane is there anything else needed to merge this?

**Comment by nirbheek** [comment]
I think this should go in.

PS: When the CI fails like that, it's best to re-push so the CI runs again.

**Comment by martingkelly** [comment]
Rebased and repushed to force the CI to run again. However, I think the AppVeyor failure is an infrastructure problem because it fails every time, so I suspect it will fail again.

Anything left to get this merged?

**Comment by martingkelly** [comment]
OK, finally CI is clean. Anything left to merge this? @nirbheek @jpakkane

**Comment by jpakkane** [comment]
Needs to be rebased to fix the :crossed_swords: conflict, but once that is done we can merge this.

**Comment by martingkelly** [comment]
OK, rebased again, should be ready to go.

**Comment by martingkelly** [comment]
(ping)

**Comment by jpakkane** [comment]
Thanks.

**Pull request merged**